### PR TITLE
Replace DateTime.UtcNow.ToString with a zero-allocation version

### PR DESCRIPTION
**Pull request by victlu**
Optimize Self-Diagnostics module to be zero-allocation #1621 

Refactor DateTime.UtcNow.ToString call in SelfDiagnosticsEventListener.WriteEvent to be allocation free.

### CODE REVIEW

**Comment by CodeBlanch** [comment]
@victlu Thanks for this! Would you mind adding some tests for the new logic?

**Comment by victlu** [comment]
> @victlu Thanks for this! Would you mind adding some tests for the new logic?

I would love to.  Can you point me in a direction of where to add this?  Seeing this is currently a private function inside an internal class.  Do we do unit tests for non-public interfaces?  I suppose I can refactor this function into some public static helper class?

**Comment by CodeBlanch** [comment]
@victlu Ya we're using this [InternalsVisibleTo](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.compilerservices.internalsvisibletoattribute) ~~hack~~ trick for unit tests. If you make it `internal` you should be able to write tests targeting it in the OpenTelemetry.Tests project:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/01dc04710adf06312e60b2311064af46560b5076/src/OpenTelemetry/AssemblyInfo.cs#L24

**Comment by victlu** [comment]
Ping

**Review by CodeBlanch — APPROVED** [review]
LGTM

**Comment by CodeBlanch** [comment]
@victlu Looks like there is a conflict now. Would you mind taking a look at that? Sorry for not getting it in sooner.

File: src/OpenTelemetry/Internal/SelfDiagnosticsEventListener.cs [hunk 1]
```diff
@@ -172,6 +171,111 @@ internal void WriteEvent(string eventMessage, ReadOnlyCollection<object> payload
...
+        /// <param name="bytes">Array of bytes to write.</param>
+        /// <param name="byteIndex">Starting index into bytes array.</param>
+        /// <returns>The number of bytes written.</returns>
+        internal int DateTimeGetBytes(DateTime datetime, byte[] bytes, int byteIndex)
+        {
+            int num;
+            int pos = byteIndex;
+
+            num = datetime.Year;
+            bytes[pos++] = (byte)('0' + ((num / 1000) % 10));
```
Comment (reyang) [inline src/OpenTelemetry/Internal/SelfDiagnosticsEventListener.cs:202]: Curious if there is a way to leverage fast BCD conversion. Is .NET doing the same thing?

Comment (reyang) [reply src/OpenTelemetry/Internal/SelfDiagnosticsEventListener.cs:202]: .NET might have some in-place formatting method that we can use - e.g. https://docs.microsoft.com/dotnet/api/system.datetime.tryformat?view=net-5.0 (this one is using `Span<T>` which won't work for net452 though).

**Review by reyang — APPROVED** [review]
LGTM.

**Pull request merged**